### PR TITLE
Add `TaprootWitness`, new methods to `XOnlyPubKey`

### PR DESCRIPTION
**Pull request by Christewart**
Pulls `TaprootWitness`, and various other methods from #3769 with the hope that this will make #3769 easier to review when it is time to merge.

There are many places in the codebase where we had pattern match exhaustiveness errors, I will go back and correct these when I start to implement signing. For now i'm just trying to implement verification of a taproot output.

The methods tested in `TaprootWitnessTest` are tested against a special branch of bitcoin core where i print out bitcoin core's results, so I'm pretty certain they are correct: https://github.com/christewart/bitcoin/tree/2022-06-12-taproot-sig-serialization

The hard coded values in `TaprootWitnessTest` are taken from the `script_assets.json`, namely this test case since it is verifying a tapscript script.

```
[  {"tx": "26dc279d02d8b1a203b653fc4e0f27f408432f3f540136d33f8f930eaeba655910095142980402000000fd697cd4eb5278f1e34545cd57b6670df806fa3a0a064fd8e385a19f1a53d9ce8d8971a30f02000000378d5fb502335dbe02000000001976a9140053a23441c8478caac4c6b769c51f8476cd4b4b88ac58020000000000001976a914f2aae94a43e0d173354201d7832b46c5269c8a2488ac4a08671e", "prevouts": ["91ca4c010000000017a9145658b58602cdf7b7e962cfe44e024cb0e366f27087", "cb127401000000002251201ebe8b90363bd097aa9f352c8b21914e1886bc09fe9e70c09f33ef2d2abdf4bc"], "index": 1,

  "success": {"scriptSig": "", "witness": ["9675a9982c6398ea9d441cb7a943bcd6ff033cc3a2e01a0178a7d3be4575be863871c6bf3eef5ecd34721c784259385ca9101c3a313e010ac942c99de05aaaa602",
  "5799cf4b193b730fb99580b186f7477c2cca4d28957326f6f1a5d14116438530e7ec0ce1cd465ad96968ae8a6a09d4d37a060a115919f56fcfebe7b2277cc2df5cc08fb6cda9105ee2512b2e22635aba",
  "7520c7b5db9562078049719228db2ac80cb9643ec96c8055aa3b29c2c03d4d99edb0ac",
  "c1a7957acbaaf7b444c53d9e0c9436e8a8a3247fd515095d66ddf6201918b40a3668f9a4ccdffcf778da624dca2dda0b08e763ec52fd4ad403ec7563a3504d0cc168b9a77a410029e01dac89567c9b2e6cd726e840351df3f2f58fefe976200a19244150d04153909f660184d656ee95fa7bf8e1d4ec83da1fca34f64bc279b76d257ec623e08baba2cfa4ea9e99646e88f1eb1668c00c0f15b7443c8ab83481611cc3ae85eb89a7bfc40067eb1d2e6354a32426d0ce710e88bc4cc0718b99c325509c9d02a6a980d675a8969be10ee9bef82cafee2fc913475667ccda37b1bc7f13f64e56c449c532658ba8481631c02ead979754c809584a875951619cec8fb040c33f06468ae0266cd8693d6a64cea5912be32d8de95a6da6300b0c50fdcd6001ea41126e7b7e5280d455054a816560028f5ca53c9a50ee52f10e15c5337315bad1f5277acb109a1418649dc6ead2fe14699742fee7182f2f15e54279c7d932ed2799d01d73c97e68bbc94d6f7f56ee0a80efd7c76e3169e10d1a1ba3b5f1eb02369dc43af687461c7a2a3344d13eb5485dca29a67f16b4cb988923060fd3b65d0f0352bb634bcc44f2fe668836dcd0f604150049835135dc4b4fbf90fb334b3938a1f137eb32f047c65b85e6c1173b890b6d0162b48b186d1f1af8521945924ac8ac8efec321bf34f1d4b3d4a304a10313052c652d53f6ecb8a55586614e8950cde9ab6fe8e22802e93b3b9139112250b80ebc589aba231af535bb20f7eeec2e412f698c17f3fdc0a2e20924a5e38b21a628a9e3b2a61e35958e60c7f5087c"]},

  "flags": "P2SH,DERSIG,CHECKLOCKTIMEVERIFY,CHECKSEQUENCEVERIFY,WITNESS,NULLDUMMY,TAPROOT",
  "final": true, "comment": "tapscript/input80limit"}
```

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala [hunk 1]
```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
...
+      //see: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#script-validation-rules
+      ControlBlock.fromBytes(stack(1))
+    } else {
+      ControlBlock.fromBytes(stack.head)
+    }
+  }
+
+  //  require(
+  //    controlBlock.length >= 33 && controlBlock.length < 33 + (32 * 128),
+  //    s"Control block length is wrong, got=${controlBlock.length}, max=${33 + (32 * 128)}")
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala]: should this be brought back?

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala [hunk 2]
```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
...
+
+  /** Call the second-to-last stack element s, the script
+    */
+  def script: ScriptPubKey = {
+    annexOpt match {
+      case Some(_) =>
+        ScriptPubKey.fromAsmBytes(stack(1))
+      case None =>
+        val spk = ScriptPubKey.fromAsmBytes(stack(1))
+        spk
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala]: these 2 are the same, doesn't seem right

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala [hunk 3]
```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
...
+          //If there are at least two witness elements, and the first byte of the last element is 0x50[4],
+          // this last element is called annex a[5] and is removed from the witness stack.
+          // The annex (or the lack of thereof) is always covered by the signature and contributes to transaction weight,
+          // but is otherwise ignored during taproot validation.
+          //see: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#script-validation-rules
+          stack(1)
+        } else {
+          stack.head
+        }
+      }
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala]: this is a repeat of above, can probably be factored out into a new function

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala [hunk 4]
```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
...
+  /** Computes the hash of the script that is a leaf in the tapscript tree
+    * @param leafVersion
+    * @param spk
+    * @see https://github.com/bitcoin/bitcoin/blob/37633d2f61697fc719390767aae740ece978b074/src/script/interpreter.cpp#L1828
+    * @return
+    */
+  def computeTapleafHash(leafVersion: Byte, spk: ScriptPubKey): Sha256Digest = {
+    val bytes =
+      ByteVector.fromByte(leafVersion) ++ spk.bytes
+    CryptoUtil.taggedSha256(bytes, "TapLeaf")
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala:376]: We should precompute the tag in crypto util to speed up validation

Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala:376]: Can be done later

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala [hunk 5]
```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
...
+      } else {
+        k = hashTapBranch(node ++ k.bytes)
+      }
+      i += 1
+    }
+    k
+  }
+
+  private def hashTapBranch(bytes: ByteVector): Sha256Digest = {
+    CryptoUtil.taggedSha256(bytes, "TapBranch")
```
Comment (benthecarman) [inline core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala:413]: ditto

File: crypto/src/main/scala/org/bitcoins/crypto/XOnlyPubKey.scala [hunk 6]
```diff
@@ -22,7 +22,49 @@ case class XOnlyPubKey(bytes: ByteVector) extends NetworkElement {
...
+        tapTweakHash(bytes ++ merkleRoot.bytes)
+      case None =>
+        // We have no scripts. The actual tweak does not matter, but follow BIP341 here to
+        // allow for reproducible tweaking.
+        tapTweakHash(bytes)
+    }
+  }
+
+  private def tapTweakHash(bytes: ByteVector): Sha256Digest = {
+    CryptoUtil.taggedSha256(bytes, "TapTweak")
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/XOnlyPubKey.scala:66]: ditto

File: crypto/src/main/scala/org/bitcoins/crypto/XOnlyPubKey.scala [hunk 7]
```diff
@@ -22,7 +22,49 @@ case class XOnlyPubKey(bytes: ByteVector) extends NetworkElement {
...
   def coord: CurveCoordinate = CurveCoordinate(bytes)
+
+  /** @see https://github.com/bitcoin/bitcoin/blob/3340d46cd363e568ce842b2a9930e30902d150ca/src/pubkey.cpp#L227
+    * @see https://github.com/bitcoin-core/secp256k1/blob/9a5a87e0f1276e0284446af1172056ea4693737f/src/modules/extrakeys/main_impl.h#L151
+    * @param pubKey the internal tapscript pubkey
+    * @param merkleRootOpt the merkle root of the tapscript tree, if empty means we have no scripts in the tapscript tree
+    * @param parity the expected parity of the public key reproduced
+    * @return
+    */
+  def checkTapTweak(
```
Comment (benthecarman) [inline crypto/src/main/scala/org/bitcoins/crypto/XOnlyPubKey.scala:36]: these functions feel like they belong somewhere, not on `XOnlyPubKey`

**Review by benthecarman — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala (hunk 1 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala]: this is now checked in the `TaprootScriptPath.isValid()` method now.

File: crypto/src/main/scala/org/bitcoins/crypto/XOnlyPubKey.scala (hunk 7 above)
Comment (Christewart) [reply crypto/src/main/scala/org/bitcoins/crypto/XOnlyPubKey.scala:36]: This is where it is in bitcoin core: 

https://github.com/bitcoin/bitcoin/blob/df933596e7e9aa17f7e5cd6e1c850520f5b56f1b/src/pubkey.cpp#L227

File: core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala (hunk 4 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala:376]: i think i'll do it later as i would rather spend the time atm on getting the script of tapscript verification done

**Review by benthecarman — APPROVED** [review]

**Pull request merged**